src/llamafactory/data/processors/supervised.py

```diff
@@ -44,22 +44,34 @@ def _encode_supervised_example(
     cutoff_len: int,
     train_on_prompt: bool,
     mask_history: bool,
+    pack_data_preprocess: bool = False,
 ) -> Tuple[List[int], List[int]]:
     messages = template.mm_plugin.process_messages(prompt + response, images, videos, processor)
     input_ids, labels = template.mm_plugin.process_token_ids([], [], images, videos, tokenizer, processor)
     encoded_pairs = template.encode_multiturn(tokenizer, messages, system, tools)
-    total_length = len(input_ids) + (1 if template.efficient_eos else 0)
+    eos_indice = len(input_ids) + (1 if template.efficient_eos else 0)
     if mask_history:
         encoded_pairs = encoded_pairs[::-1]  # high priority for last turns
 
     for turn_idx, (source_ids, target_ids) in enumerate(encoded_pairs):
-        if total_length >= cutoff_len:
-            break
 
-        source_len, target_len = infer_seqlen(len(source_ids), len(target_ids), cutoff_len - total_length)
+        if pack_data_preprocess and len(source_ids)+len(target_ids) >= cutoff_len:
+            raise ValueError(f"""Packing dataset `len(source_ids)+len(target_ids)` needs a larger cutoff_len:
+                              {len(source_ids)+len(target_ids)}> {cutoff_len}""")
+        else:
+            if eos_indice >= cutoff_len:
+                logger.warning(f"""cutoff_len {cutoff_len} is too small for the input turn_idx: {turn_idx}, drop it.
+                            eg: The eos_indice is exactly one less than the bubble length, causing the last one to be discarded.
+                            """)
+                break
```

Comment:
L59 raise exception.
curious why L66 just break?

Reply:
When pack_data_preprocess is true, cutoff_len is not used for truncating the input

`pack_data_preprocess and len(source_ids)+len(target_ids) >= cutoff_len:`
Used for verifying the maximum packing of long texts. For example, when the message length is >= 21, it should report an error instead of discarding the data if it doesn't form a complete training pack.

Ideal situation:
![image](https://github.com/user-attachments/assets/a5a3a635-e280-4649-ab23-6e36f00f3c89)

Cases where an error should be reported:
![image](https://github.com/user-attachments/assets/ab4e264e-b243-4e5d-875f-c93deba66739)

preprocess_packed_supervised_dataset receives batched data from dataset.map. When the number of processing threads is 1, only one process handles the data. The graph is too abstract; normally, it would be divided into batch_size pieces for all processes to handle.

@hiyouga